### PR TITLE
EBP-19: As a Solace Developer, I can use cache containers in integration testing

### PR DESCRIPTION
**Pull request by TrentDaniel**
This PR includes the changes necessary to create cache instances, configure the broker to communicate with them, and use these cache instances during cache integration testing.
- [x] The standard ginkgo command in an unconfigured environment does not cause cache tests to be run, and does not cause cache containers or configuration to be run
- [x] The standard ginkgo command in an environment configured with the SOLCACHE_TEST_IMAGE variable set to any value will run the cache tests and run cache containers and configurations
- [x] Running a specific test or group of tests that do not include cache does not cause cache containers or broker configurations for cache to be run
- [x] Running a specific test or a group of tests that include cache in at least one test cause cache containers or configurations to be run
- [x] The tests can create their own cache containers as a part of starting the test, run any configurations, and destroy those containers after the test has completed
- [x] The name of the cache docker images as well as the docker registry where they are stored must not be committed to the repo
- [x] The tests must have access to a solcacheproxy image to run those tests which require solcache proxy, and access to a solcache image to run those tests which require solcache. The env var SOLCACHEPROXY_TEST_IMAGE should be used in a similar way to SOLCACHE_IMAGE.
- [x] The tests must be able to configure the broker for cache over SEMP
- [x] A README must be provided to instruct the user on how to setup and run cache tests. This README must also communicate that the cache and cache proxy images are proprietary, and they will need to purchase them from the company to run the tests.
- [x] All cache SEMP calls should be verified to work through manual inspection of the broker, and any notes on usage of these semp calls should be documented in a README for future developers to use.
- [x] Verify that a cache request/response can be sent through the broker using sdkperf after the go test infrastructure has configured the broker for cache
- [x] Verify that the broker configuration is consistent with the Python API broker configuration

### CODE REVIEW

File: test/README.md [hunk 1]
```diff
@@ -41,6 +41,20 @@ Kerberos tests can be run when using testcontainers by exporting the variable KR
 ### OAuth
 OAuth tests can be run when using testcontainers by exporting the variable OAUTH_TEST_IMAGE with a valid OAuth server image.
 
+### Cache
+Cache tests can be run when using testcaontainers by exporting the variables:
+- PUBSUB_CACHE_HOSTNAME=<name of cache instance>
+- PUBSUB_CACHE_SUPSECT_HOSTNAME=<name of suspect cache instance>
+- PUBSUB_CACHE_TEST_IMAGE=<name of docker image for cache instances>
+- PUBSUB_CACHEPROXY_TEST_IMAGE=<name of docker image for cahce proxy instances>
+These environment variables are used to create and destroy the various docker containers that are required for testing. Modifying these variables after they have been used to create containers will lead to undefined behaviour and should not be done. Populating these variables after executing a test command should not be expected to cause the containers to be created partway through the tests, and will lead to undefined behaviour.
```
Comment (TrentDaniel) [inline test/README.md:50]: change 'after test execution' with 'during execution'

File: test/README.md [hunk 2]
```diff
@@ -41,6 +41,20 @@ Kerberos tests can be run when using testcontainers by exporting the variable KR
 ### OAuth
 OAuth tests can be run when using testcontainers by exporting the variable OAUTH_TEST_IMAGE with a valid OAuth server image.
 
+### Cache
+Cache tests can be run when using testcaontainers by exporting the variables:
```
Comment (TrentDaniel) [inline test/README.md:45]: fix typo

File: test/README.md [hunk 3]
```diff
@@ -41,6 +41,20 @@ Kerberos tests can be run when using testcontainers by exporting the variable KR
...
+- PUBSUB_CACHE_HOSTNAME=<name of cache instance>
+- PUBSUB_CACHE_SUPSECT_HOSTNAME=<name of suspect cache instance>
+- PUBSUB_CACHE_TEST_IMAGE=<name of docker image for cache instances>
+- PUBSUB_CACHEPROXY_TEST_IMAGE=<name of docker image for cahce proxy instances>
+These environment variables are used to create and destroy the various docker containers that are required for testing. Modifying these variables after they have been used to create containers will lead to undefined behaviour and should not be done. Populating these variables after executing a test command should not be expected to cause the containers to be created partway through the tests, and will lead to undefined behaviour.
+PUBSUB_CACHE_HOSTNAME, PUBSUB_CACHE_SUPSECT_HOSTNAME, and PUBSUB_CACHE_TEST_IMAGE are all necessary to run cache tests.
+PUBSUB_CACHE_HOSTNAME, and PUBSUB_CACHEPROXY_TEST_IMAGE are all necessary to run cache proxy tests. For tests that use both cache and cache proxy, all four variables are needed. Cache suspect is just a specfially configured cache instance, so tests that use suspect cache have the same requirements as those that use cache.
```
Comment (TrentDaniel) [inline test/README.md:52]: Make sure there are defaults for cache host name and suspect host name

File: test/cache_test.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+var _ = Describe("Cache Strategy", func() {
+        /* The following tests are just demos of how to check for infrastructure.
+        They should be removed once proper integration tests are added. */
+    Describe("Demo isolated Cache functionality", func() {
+        BeforeEach(func () {
+            CheckCache() // skips test with message if cache image is not available
+        })
+        It("should run this test", func() {
+                fmt.Println("Running cache test")
```
Comment (TrentDaniel) [inline test/cache_test.go:47]: Remove println, add output as assertion failure instead

File: test/helpers/semp_helpers.go [hunk 5]
```diff
@@ -23,18 +23,9 @@ import (
...
-var True = boolPointer(true)
-
-// False defined
-var False = boolPointer(false)
-
-func boolPointer(b bool) **bool {
-	bp := &b
-	return &bp
-}
+var True = testcontext.True
```
Comment (TrentDaniel) [inline test/helpers/semp_helpers.go:26]: Duplicate comment

File: test/testcontext/cache_util.go [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+
+import (
+	"time"
+	"net/url"
+    "fmt"
+
+	"solace.dev/go/messaging/test/sempclient/config"
+)
+
+func ConfigureDistributedCache(semp SEMPv2, msgVpnName string, distributedCache DistributedCacheConfig) error {
```
Comment (TrentDaniel) [inline test/testcontext/cache_util.go:27]: Change name to ConfigureDistributedCacheIfNotExists

File: test/testcontext/cache_util.go [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+        }
+
+    return err
+}
+
+// waitForOperationalCache checks that each cache instance is in the expected operational state. This has only been
+// tested when checking for an operational state of CacheOperationalStateUp, where it is necessary that if all cache
+// instances are up, then all distributed caches and cache clusters are also up. This relationship may not hold for
+// other operational states. The usage of this method for checking other operational states may require refactoring.
+func (ctx *testContextCommon) waitForOperationalCache(timeout float64, state CacheOperationalState) error {
```
Comment (TrentDaniel) [inline test/testcontext/cache_util.go:251]: Pass time.Duration instead of float64

**Review by cjwmorgan-sol — COMMENTED** [review]
Did review together found minor issues please resolve comments.
Once jeknins autoamtion passes wiht changes this should be ready for merge.

File: test/testcontext/test_context_testcontainers.go [hunk 8]
```diff
@@ -52,6 +56,11 @@ func (context *testContainersTestContext) Setup() error {
 	fmt.Println()
 	fmt.Println("-- TESTCONTAINERS SETUP --")
 
+    testFolderVar := os.Getenv(testFolderVarName)
```
Comment (TrentDaniel) [inline test/testcontext/test_context_testcontainers.go:59]: Add comment explaining this

**Review by cjwmorgan-sol — CHANGES_REQUESTED** [review]
Needs one last change. To make the test code use more golang and avoid "python" like code.

**Review by cjwmorgan-sol — APPROVED** [review]
Changes look good to me.
Git actions are passing.

**Pull request merged**